mam/v2/api/api.c

```diff
@@ -183,12 +198,112 @@ retcode_t mam_api_bundle_write_packet(mam_api_t *const api,
+  iota_transaction_t *curr_tx = (iota_transaction_t *)utarray_eltptr(bundle, 0);
+
+  trit_t *msg_trits_ptr = msg_trits;
+  size_t curr_tx_index = 0;
+  BUNDLE_FOREACH(bundle, curr_tx) {
+    if (curr_tx_index++ < start_index) {
+      continue;
+    }
+    flex_trits_to_trits(msg_trits_ptr, NUM_TRITS_SIGNATURE,
+                        transaction_signature(curr_tx), NUM_TRITS_SIGNATURE,
+                        NUM_TRITS_SIGNATURE);
+    msg_trits_ptr += NUM_TRITS_SIGNATURE;
+  }
+
+  return trits_from_rep(num_trits_in_bundle, msg_trits);
+}
+
+static retcode_t mam_api_bundle_read_header(
+    mam_msg_recv_context_t const *const ctx, trits_t *const msg) {
+  retcode_t err;
+  ERR_BIND_RETURN(mam_msg_recv(ctx, msg), err);
+  return err;
+}
+
+retcode_t mam_api_bundle_read_msg(mam_api_t *const api,
+                                  bundle_transactions_t const *const bundle,
+                                  flex_trit_t **const packet_payload) {
+  retcode_t err = RC_OK;
+  MAM2_ASSERT(packet_payload && *packet_payload == NULL);
+  if (!mam_api_bundle_contains_header(bundle)) {
+    return RC_MAM2_BUNDLE_DOES_NOT_CONTAIN_HEADER;
+  }
+
+  mam_msg_recv_context_t cfg;
+  cfg.pubkey = -1;
+  cfg.psks = api->ntru_pks;
+  cfg.ntrus = api->ntru_sks;
+  trit_t ch_id_trits_raw[NUM_TRITS_ADDRESS];
+  iota_transaction_t *curr_tx = (iota_transaction_t *)utarray_eltptr(bundle, 0);
+  flex_trits_to_trits(ch_id_trits_raw, NUM_TRITS_ADDRESS,
+                      transaction_address(curr_tx), NUM_TRITS_ADDRESS,
+                      NUM_TRITS_ADDRESS);
+  trits_t ch_id_trits = trits_from_rep(NUM_TRITS_ADDRESS, ch_id_trits_raw);
+  trits_copy(ch_id_trits, mam_msg_recv_cfg_chid(&cfg));
+
+  // Flatten flex_trits encoded in transaction sig_or_fragment field
+  // into a single long trits_t data structure
+  size_t num_trits_in_bundle =
+      bundle_transactions_size(bundle) * NUM_TRITS_SIGNATURE;
+  trit_t msg_trits[num_trits_in_bundle];
+  trits_t msg =
+      msg_trits_from_bundle(bundle, msg_trits, num_trits_in_bundle, 0);
+
+  ERR_BIND_RETURN(mam_api_bundle_read_header(&cfg, &msg), err);
+
+  ERR_BIND_RETURN(trit_t_to_mam_msg_recv_context_t_map_add(
+                      &api->recv_ctxs, mam_msg_recv_cfg_msg_id(&cfg).p, cfg),
+                  err);
+
+  size_t packet_index = msg.d / NUM_TRITS_SIGNATURE + 1;
+  if (packet_index < bundle_transactions_size(bundle)) {
+    ERR_BIND_RETURN(mam_api_bundle_read_packet(api, bundle, packet_payload,
+                                               mam_msg_recv_cfg_msg_id(&cfg).p,
+                                               packet_index),
+                    err);
   }
+
   return RC_OK;
 }
+
+retcode_t mam_api_bundle_read_packet(mam_api_t const *const api,
+                                     bundle_transactions_t const *const bundle,
+                                     flex_trit_t **const packet_payload,
+                                     trit_t const *const msg_id,
```

Comment:
msg_id is not needed since it's on the tag. Otherwise we have to ask the user to provide it.